modules/dxp/apps/analytics/analytics-reports-web/src/main/resources/META-INF/resources/js/components/Translation.js

```diff
@@ -0,0 +1,138 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * The contents of this file are subject to the terms of the Liferay Enterprise
+ * Subscription License ("License"). You may not use this file except in
+ * compliance with the License. You can obtain a copy of the License by
+ * contacting Liferay, Inc. See the License for the specific language governing
+ * permissions and limitations under the License, including but not limited to
+ * distribution rights of the Software.
+ */
+
+import ClayButton from '@clayui/button';
+import ClayDropDown from '@clayui/drop-down';
+import ClayIcon from '@clayui/icon';
+import ClayLabel from '@clayui/label';
+import ClayLayout from '@clayui/layout';
+import PropTypes from 'prop-types';
+import React, {useMemo, useState} from 'react';
+
+import {useChartState} from '../state/chartState';
+
+export default function Translation({
+	defaultLanguage,
+	onSelectedLanguageClick,
+	publishDate,
+	timeSpanKey,
+	viewURLs,
+}) {
+	const [active, setActive] = useState(false);
+
+	const selectedLanguage = useMemo(() => {
+		return (
+			viewURLs.find((language) => language.selected).languageId ||
+			defaultLanguage
+		);
+	}, [defaultLanguage, viewURLs]);
+
+	const {state: chartState} = useChartState({
+		publishDate,
+		timeSpanKey,
+	});
+
+	const timeSpanOffset = useMemo(() => {
```

Comment:
I'll change it 👍

Reply:
One little thing about both `useMemo` and `useCallback` to bear in mind is that, in addition to the use case that @carloslancha is talking about (memoizing expensive calculations), there is one other important one:

**Preserving the identity of values across renders.**

For example, imagine I have an `onChange` callback that I am passing down to a child component. According to the "memoize expensive calculation" heuristic, I might not want to use `useCallback`, because defining a function is super cheap.

But, depending on the contents of the child, maybe the child ends up having to pass that callback into the dependencies array of a `useEffect` hook, and maybe _that_ ends up being a significant problem. If the identity of the callback changes, the effect will run again, and the effects could be problematic, ranging from a minor performance issue like an unnecessary re-render through to something quite bad like a duplicate network operation. (Worse still if the network operation _writes_ to the database.)

There is no blanket rule here (ie. "always use `useCallback`/`useMemo` when passing values to children) because it would be quite wasteful if the child itself didn't do anything that would benefit from having stable values over time. So you really need to make the decision based on knowledge of the context. And as a rule of thumb, the more generic/general your component is (ie. the more likely it is to be composed with other components that you don't have intimate knowledge of), the more you should err on the side of caution when it comes to making a best effort to guarantee stability of values.

Reply:
Thx @wincent !